### PR TITLE
Covidsim kernel

### PR DESCRIPTION
**Pull request by Elizabeth-Hayman**
Quick addition to add the covidsim gravity kernel to apply to the spatial sweep

### CODE REVIEW

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#
+# Class of Covidsim-style gravity kernels
+#
+
+
+class Kernel:
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:6]: I think this is only a spatial Kernel right? Might be worth calling it `SpatialKernel` to avoid confusion

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    """Class to create the gravity kernel used throughout
+    CovidSim in the spatial weighting.
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:8]: Might be worth citing it here, and certainly adding to the wiki with reference!

File: pyEpiabm/pyEpiabm/sweep/spatial_sweep.py [hunk 3]
```diff
@@ -195,8 +196,9 @@ def find_infectees_Covidsim(self, infector: Person,
         while number_to_infect > 0 and count < self._population.total_people():
             count += 1
             # Weighting for cell choice in Covidsim uses cum_trans and
-            # invCDF arrays, but really can't see how these are
-            # initialised. Have used the number of susceptible for now.
+            # invCDF arrays, which are equivilent to weighting by total
```
Comment (rcw5890) [inline pyEpiabm/pyEpiabm/sweep/spatial_sweep.py:199]:
```suggestion
            # invCDF arrays, which are equivalent to weighting by total
```

File: pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+numReps = 1
+
+
+class TestKernel(unittest.TestCase):
+    """Test the 'Kernel' class.
+    """
+    @parameterized.expand([(np.random.rand(1) * numReps,)
+                          for _ in range(numReps)])
+    def test_kernel(self, dist):
+        dist = np.sort(dist)
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py:16]:
```suggestion
```
Not necessary on array of length one

File: pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        dist = 10 * dist[0]
+        dist_object = Kernel()
+        value = dist_object.weighting(dist)
+        print(value)
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py:20]:
```suggestion
```

File: pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import unittest
+import numpy as np
+from parameterized import parameterized
+
+from pyEpiabm.utility import Kernel
+
+numReps = 1
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py:7]: Can this be increased without significant runtime cost?

File: pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        dist = 10 * dist[0]
+        dist_object = Kernel()
+        value = dist_object.weighting(dist)
+        print(value)
+        self.assertTrue(0 <= value)
+
+    @parameterized.expand([(np.random.rand(3) * numReps,)
+                           for _ in range(numReps)])
+    def test_kernel_shape_size(self, dist):
+        dist = np.sort(dist)
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py:26]:
```suggestion
```

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    """
+    @staticmethod
+    def weighting(distance: float, scale: float = 4, shape: float = 3.8):
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:11]: Can these values be read in from the parameters file

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+            Distance input as the main argument
+        scale : Float
+            Parameter to scale the kernel function
+        shape : Float
+            Parameter to change the shape of the kernel function
+
+        Returns
+        ----------
+        weight : Float
+            Float of the weight function
```
Comment (rcw5890) [inline pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:27]: :star_struck: the new docstring format

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+
+class Kernel:
+    """Class to create the gravity kernel used throughout
+    CovidSim in the spatial weighting.
+    """
+    @staticmethod
+    def weighting(distance: float, scale: float = 4, shape: float = 3.8):
+        """Returns the weighting given by the Covidsim distance
+        weighting. Default parameters use the Covidsim GB spatial data.
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:13]: I think these are commuting data for that kernel rather than spatial data

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+        Parameters
+        ----------
+        distance : Float
```
Comment (rcw5890) [inline pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:17]:
```suggestion
        distance : float
```
&c

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py (hunk 9 above)
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:27]:
```suggestion
        distance : float
            Distance input as the main argument
        scale : float
            Parameter to scale the kernel function
        shape : float
            Parameter to change the shape of the kernel function

        Returns
        ----------
        weight : float
            Float of the weight function
```

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+            Parameter to change the shape of the kernel function
+
+        Returns
+        ----------
+        weight : Float
+            Float of the weight function
+
+        """
+
+        return 1/(1+distance/scale)**shape
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:31]:
```suggestion
        return 1 / (1 + distance / scale) ** shape
```
Spaces around operators in python

File: pyEpiabm/pyEpiabm/sweep/spatial_sweep.py [hunk 13]
```diff
@@ -205,8 +207,11 @@ def find_infectees_Covidsim(self, infector: Person,
             # Sample at random from the infectee cell to find
             # an infectee
             infectee = random.sample(infectee_cell.persons, 1)[0]
-            infection_distance = (DistanceFunctions.dist(
-                infector.microcell.cell.location, infectee_cell.location) /
+            # Covidsim used explicitly the kernel distance between people,
+            # and divided by the kernel of the minimum distance between
+            # their cells.
```
Comment (rcw5890) [inline pyEpiabm/pyEpiabm/sweep/spatial_sweep.py:212]: I would clarify this comment and what you mean by minimum distance

File: pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py (hunk 5 above)
Comment (rcw5890) [inline pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py:20]: Do we still need this

File: pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+        dist = 10 * dist
+        dist_object = Kernel()
+        value = dist_object.weighting(dist[0], dist[1], dist[2])
+        self.assertTrue(0 <= value)
```
Comment (rcw5890) [inline pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py:30]: You could also test that the value is close to what you expect, not just that it is greater than zero

File: pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py:31]: Can we have a few extra tests:
* Check exact value of kernel is correct for specified parameter values (can be simple numbers)
* Check that kernel weighting decreases for larger separations

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py (hunk 10 above)
Comment (rcw5890) [inline pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:13]: Its not clear how the parameters are related to spatial data. If it's relevant, I would elaborate this docstring.

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+        ----------
+        distance : Float
+            Distance input as the main argument
+        scale : Float
+            Parameter to scale the kernel function
+        shape : Float
+            Parameter to change the shape of the kernel function
+
+        Returns
+        ----------
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:25]:
```suggestion
        -------
```
Should be aligned with `Returns` above

**Review by KCGallagher — CHANGES_REQUESTED** [review]
Looks good but might need a little more time to be complete!

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py (hunk 10 above)
Comment (rcw5890) [inline pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:13]:
```suggestion
        weighting. Default parameters use the Covidsim GB spatial data.
        
        The formula is given by
        
        .. math::
            \frac{1}{\left(1+\frac{\text{distance}}{\text{scale}}\right)^\text{shape}}
```
While it is easy enough to read from the source code in this case, it doesn't hurt to give it in the docstring, and I'm also curious whether we have the latex working on readthedocs and sphinx.

**Review by rcw5890 — CHANGES_REQUESTED** [review]

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py (hunk 10 above)
Comment (Elizabeth-Hayman) [reply pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:13]: see above

File: pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py (hunk 5 above)
Comment (Elizabeth-Hayman) [reply pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py:20]: nope!

File: pyEpiabm/pyEpiabm/sweep/spatial_sweep.py (hunk 13 above)
Comment (Elizabeth-Hayman) [reply pyEpiabm/pyEpiabm/sweep/spatial_sweep.py:212]: Not sure if my rewrite helped, let me know if you want it further explained

File: pyEpiabm/pyEpiabm/sweep/spatial_sweep.py [hunk 17]
```diff
@@ -12,6 +12,7 @@
 from pyEpiabm.property import InfectionStatus
 from pyEpiabm.routine import SpatialInfection
 from pyEpiabm.utility import DistanceFunctions
+from pyEpiabm.utility import SpatialKernel as SK
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/sweep/spatial_sweep.py:15]: Is this really necessary? We don't use this syntax anywhere else, and PEP is dubious on it...

File: pyEpiabm/pyEpiabm/sweep/spatial_sweep.py [hunk 18]
```diff
@@ -195,8 +196,9 @@ def find_infectees_Covidsim(self, infector: Person,
         while number_to_infect > 0 and count < self._population.total_people():
             count += 1
             # Weighting for cell choice in Covidsim uses cum_trans and
-            # invCDF arrays, but really can't see how these are
-            # initialised. Have used the number of susceptible for now.
+            # invCDF arrays, which are equivalent to weighting by total
+            # susceptibles*max_transmission. May want to add transmission
+            # parameter later.
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/sweep/spatial_sweep.py:201]: What transmission parameter out of interest? I am currently adding an equivalent of `P.HouseholdTrans` from CovidSim, and intned to use equivalents for place etc

File: pyEpiabm/pyEpiabm/sweep/spatial_sweep.py (hunk 17 above)
Comment (KCGallagher) [reply pyEpiabm/pyEpiabm/sweep/spatial_sweep.py:15]: (If it is to shorten L214, just calculate the distance in a separate line!)

File: pyEpiabm/pyEpiabm/sweep/spatial_sweep.py [hunk 19]
```diff
@@ -205,10 +207,14 @@ def find_infectees_Covidsim(self, infector: Person,
...
-                infector.microcell.cell.location, infectee_cell.location) /
+            # Covidsim tested each infection event by testing the ratio
+            # of the spatial kernel applied to the distance between people
+            # to the spatial kernel of the shorted distance between
+            # their cells.
+            infection_distance = (SK.weighting(DistanceFunctions.dist(
+                infector.microcell.cell.location, infectee_cell.location)) /
                 Parameters.instance().infection_radius)
-            if (infection_distance < random.random()):
+            if (infection_distance > random.random()):
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/sweep/spatial_sweep.py:217]: Just to check, is this the right way round - seems to be adding people when the infection distance is large?

File: pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+class TestSpatialKernel(unittest.TestCase):
+    """Test the 'SpatialKernel' class.
+    """
+    @parameterized.expand([(np.random.rand(1) * numReps,)
+                          for _ in range(numReps)])
+    def test_kernel(self, dist):
+        dist = 10 * dist[0]
+        dist_object = SpatialKernel()
+        value = dist_object.weighting(dist)
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py:18]: Could just pass in dist[0] here - not sure the need for multiplying by 10, and given our default grid is 0 to 1 the `np.random.rand` is rather apt actually?

File: pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+        self.assertTrue(0 <= value)
+
+    @parameterized.expand([(np.random.rand(3) * numReps,)
+                           for _ in range(numReps)])
+    def test_kernel_shape_size(self, dist):
+        dist = 10 * dist
+        dist_object = SpatialKernel()
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py:25]: Do we need to create an instance of the class if we are only calling a static method?

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py (hunk 2 above)
Comment (KCGallagher) [reply pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:8]: That now looks good - but maybe have a link to the page where the detail is (that seems to go to the homepage?)

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+
+        Parameters
+        ----------
+        distance : float
+            Distance input as the main argument
+        scale : float
+            Parameter to scale the kernel function
+        shape : float
+            Parameter to change the shape of the kernel function
+        Returns
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:30]:
```suggestion
            Parameter to change the shape of the kernel function

        Returns
```

File: pyEpiabm/pyEpiabm/utility/covidsim_kernel.py [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+        -------
+        weight : float
+            Float of the weight function
+
+        """
+        assert (scale > 0), "Spatial kernel scale must be positive."
+        assert (shape > 0), "Spatial kernel shape must be positive."
+        return 1 / (1+distance / scale) ** shape
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/utility/covidsim_kernel.py:38]:
```suggestion
        return 1 / (1 + distance / scale) ** shape
```

**Review by KCGallagher — CHANGES_REQUESTED** [review]

File: pyEpiabm/pyEpiabm/sweep/spatial_sweep.py (hunk 18 above)
Comment (Elizabeth-Hayman) [reply pyEpiabm/pyEpiabm/sweep/spatial_sweep.py:201]: max transmission does seem to be entirely dependent on the mininum distance between cells. You seem to think this makes sense. I don't.

File: pyEpiabm/pyEpiabm/sweep/spatial_sweep.py (hunk 19 above)
Comment (Elizabeth-Hayman) [reply pyEpiabm/pyEpiabm/sweep/spatial_sweep.py:217]: Yes bc the kernel is monotonic decreasing now. Renamed vriables to make this clear

File: pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py (hunk 21 above)
Comment (Elizabeth-Hayman) [reply pyEpiabm/pyEpiabm/tests/test_utility/test_covidsim_kernel.py:25]: nope

File: pyEpiabm/pyEpiabm/sweep/spatial_sweep.py [hunk 24]
```diff
@@ -203,10 +205,17 @@ def find_infectees_Covidsim(self, infector: Person,
             # Sample at random from the infectee cell to find
             # an infectee
             infectee = random.sample(infectee_cell.persons, 1)[0]
-            infection_distance = (DistanceFunctions.dist(
-                infector.microcell.cell.location, infectee_cell.location) /
-                Parameters.instance().infection_radius)
-            if (infection_distance < random.random()):
+            # Covidsim tested each infection event by testing the ratio
+            # of the spatial kernel applied to the distance between people
+            # to the spatial kernel of the shorted distance between
```
Comment (rcw5890) [inline pyEpiabm/pyEpiabm/sweep/spatial_sweep.py:210]:
```suggestion
            # to the spatial kernel of the shortest distance between
```

**Review by rcw5890 — CHANGES_REQUESTED** [review]

**Review by I-Bouros — APPROVED** [review]

File: pyEpiabm/pyEpiabm/core/microcell.py [hunk 25]
```diff
@@ -109,3 +110,17 @@ def notify_person_status_change(
         """
         self.compartment_counter.report(old_status, new_status)
         self.cell.notify_person_status_change(old_status, new_status)
+
+    def set_location(self, loc: typing.Tuple[float, float]):
+        """Method to set or change the location of a cell.
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/core/microcell.py:115]:
```suggestion
        """Method to set or change the location of a microcell.
```

File: pyEpiabm/pyEpiabm/core/microcell.py [hunk 26]
```diff
@@ -109,3 +110,17 @@ def notify_person_status_change(
...
+
+        Parameters
+        ----------
+        loc : Tuple[float, float]
+            (x,y) coordinates of the place
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/core/microcell.py:120]:
```suggestion
            (x,y) coordinates of the microcell
```

File: pyEpiabm/pyEpiabm/utility/distance_metrics.py [hunk 27]
```diff
@@ -101,3 +102,29 @@ def dist_periodic(loc1: typing.Tuple[int, int],
             if diff[index] > 0.5*scales[index]:
                 diff[index] = scales[index] - diff[index]
         return np.linalg.norm(diff)
+
+    def minimum_between_cells(cell1, cell2):
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/utility/distance_metrics.py:106]: At the moment this will result in certain infection, as we don't have a spatial microcell distribution, so the distance between microcells will be the same as the distance between cells, and so the ratio of spatial kernels will be unity. Ideally we should implement spatial distribution of microcells, but at the very least log a warning here if the minimum distance is the same as the distance between cells, so that the user is aware that microcell locations probably aren't set, and you may be getting certain infection on the spatial sweep

**Review by KCGallagher — CHANGES_REQUESTED** [review]

File: pyEpiabm/pyEpiabm/utility/distance_metrics.py (hunk 27 above)
Comment (Elizabeth-Hayman) [reply pyEpiabm/pyEpiabm/utility/distance_metrics.py:106]: After discussion will set microcells in the toy population config on the next pull request

Comment (Elizabeth-Hayman) [reply pyEpiabm/pyEpiabm/utility/distance_metrics.py:106]: Have now implemented

File: pyEpiabm/docs/source/utility.rst [hunk 28]
```diff
@@ -12,6 +12,8 @@ Overview:
 - :class:`InverseCdf`
 - :class:`RandomMethods`
 - :class:`StateTransitionMatrix`
+- :class:`SpatialKernel`
+- :class:`StateTransitionMatrix`
```
Comment (KCGallagher) [inline pyEpiabm/docs/source/utility.rst:17]:
```suggestion
```

File: pyEpiabm/pyEpiabm/routine/toy_population_config.py [hunk 29]
```diff
@@ -11,6 +11,7 @@
 from pyEpiabm.core import Household, Population
 from pyEpiabm.property import PlaceType
 from pyEpiabm.utility import log_exceptions
+from pyEpiabm.utility import DistanceFunctions
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/routine/toy_population_config.py:14]:
```suggestion
from pyEpiabm.utility import DistanceFunctions, log_exceptions
```

File: pyEpiabm/pyEpiabm/routine/toy_population_config.py [hunk 30]
```diff
@@ -168,12 +169,30 @@ def assign_cell_locations(population: Population, method: str = 'random'):
             if method == "random":
                 for cell in population.cells:
                     cell.set_location(tuple(np.random.rand(2)))
+                for cell in population.cells:
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/routine/toy_population_config.py:172]:
```suggestion
```

File: pyEpiabm/pyEpiabm/sweep/spatial_sweep.py [hunk 31]
```diff
@@ -11,6 +11,7 @@
 from pyEpiabm.core import Cell, Parameters, Person
 from pyEpiabm.property import InfectionStatus, SpatialInfection
 from pyEpiabm.utility import DistanceFunctions
+from pyEpiabm.utility import SpatialKernel
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/sweep/spatial_sweep.py:14]:
```suggestion
from pyEpiabm.utility import DistanceFunctions, SpatialKernel
```

File: pyEpiabm/pyEpiabm/tests/test_routine/test_toy_population_config.py [hunk 32]
```diff
@@ -200,6 +205,16 @@ def test_assign_cell_locations_grid(self, cell_num):
...
+                test_x = (cell.location[0] +
+                          (j % mcell_len - .5 * (mcell_len - 1)) /
+                          (grid_len * (mcell_len - 1)))
+                test_y = (cell.location[1] +
+                          (j // mcell_len - .5 * (mcell_len - 1)) /
+                          (grid_len * (mcell_len - 1)))
+                self.assertAlmostEqual(mcell.location[0], test_x)
+                self.assertAlmostEqual(mcell.location[1], test_y)
 
     def test_assign_cell_locations_known_grid(self):
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/tests/test_routine/test_toy_population_config.py:219]: Might be nice to have a cheeky microcell position check in here too?
And also a check for the random method that all microcells belong to the nearest cell (more thought but definitely good to have!)

**Review by KCGallagher — CHANGES_REQUESTED** [review]

**Review by rcw5890 — APPROVED** [review]

File: pyEpiabm/docs/source/utility.rst [hunk 33]
```diff
@@ -12,7 +12,7 @@ Overview:
 - :class:`InverseCdf`
 - :class:`RandomMethods`
 - :class:`StateTransitionMatrix`
```
Comment (KCGallagher) [inline pyEpiabm/docs/source/utility.rst:14]:
```suggestion
```

File: pyEpiabm/pyEpiabm/routine/toy_population_config.py [hunk 34]
```diff
@@ -168,12 +168,29 @@ def assign_cell_locations(population: Population, method: str = 'random'):
             if method == "random":
                 for cell in population.cells:
                     cell.set_location(tuple(np.random.rand(2)))
+                    for microcell in cell.microcells:
+                        inter_dist = [0]
+                        cell_dist = 100
+                        while min(inter_dist) < cell_dist:
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/routine/toy_population_config.py:174]: This is a little ugly given we never use these default values - a do/while loop might be more suitable here?

File: pyEpiabm/pyEpiabm/utility/distance_metrics.py [hunk 35]
```diff
@@ -101,3 +102,29 @@ def dist_periodic(loc1: typing.Tuple[int, int],
             if diff[index] > 0.5*scales[index]:
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/utility/distance_metrics.py:102]:
```suggestion
            if diff[index] > 0.5 * scales[index]:
```

**Review by KCGallagher — CHANGES_REQUESTED** [review]
This is looking really good, thanks for doing this!

File: pyEpiabm/pyEpiabm/routine/toy_population_config.py (hunk 34 above)
Comment (Elizabeth-Hayman) [reply pyEpiabm/pyEpiabm/routine/toy_population_config.py]: Minor reworking, but lack of do while in python does make this ugly!

File: pyEpiabm/pyEpiabm/routine/toy_population_config.py [hunk 36]
```diff
@@ -168,12 +168,29 @@ def assign_cell_locations(population: Population, method: str = 'random'):
...
+                            # Will keep random location only if microcell
+                            # is closer to its cell's location than any other.
+                            # Not very efficient.
+                            microcell.set_location(tuple(np.random.rand(2)))
+                            cell_dist = (DistanceFunctions.dist(microcell.
+                                         location, cell.location))
+                            inter_dist = [DistanceFunctions.dist(microcell.
+                                          location, cell2.location) for cell2
+                                          in population.cells]
+                            if not min(inter_dist) < cell_dist:
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/routine/toy_population_config.py:182]:
```suggestion
                            if min(inter_dist) == cell_dist:
```

**Review by KCGallagher — CHANGES_REQUESTED** [review]
One tiny change then I will be happy to approve!

File: pyEpiabm/pyEpiabm/routine/toy_population_config.py [hunk 37]
```diff
@@ -182,6 +199,15 @@ def assign_cell_locations(population: Population, method: str = 'random'):
...
                     cell.set_location((pos[i % grid_len],
                                        pos[i // grid_len]))
+                    mcell_num = len(cell.microcells)
+                    mcell_len = math.ceil(math.sqrt(mcell_num))
+                    m_pos = np.linspace(0, 1, mcell_len)
+                    for j, microcell in enumerate(cell.microcells):
+                        x = pos[i % grid_len] + \
+                            (m_pos[j % mcell_len]-0.5)/grid_len
+                        y = pos[i // grid_len] + \
+                            (m_pos[j // mcell_len]-0.5)/grid_len
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/routine/toy_population_config.py:209]:
```suggestion
                            (m_pos[j % mcell_len] - 0.5) / grid_len
                        y = pos[i // grid_len] + \
                            (m_pos[j // mcell_len] - 0.5) / grid_len
```

**Review by KCGallagher — CHANGES_REQUESTED** [review]

**Review by KCGallagher — APPROVED** [review]
